cmd/git/main.go

```diff
@@ -0,0 +1,316 @@
+		case typePrivateKey:
+			var sshCmd = []string{"ssh",
+				"-i",
+				filepath.Join(flagValues.secretPath, "ssh-privatekey"),
+			}
+
+			var knownHostsFile = filepath.Join(flagValues.secretPath, "known_hosts")
+			if hasFile(knownHostsFile) {
+				sshCmd = append(sshCmd,
+					"-o", "GlobalKnownHostsFile=/dev/null",
+					"-o", fmt.Sprintf("UserKnownHostsFile=%s", knownHostsFile),
+				)
+			} else {
+				sshCmd = append(sshCmd,
+					"-o", "StrictHostKeyChecking=accept-new",
+				)
+			}
+
+			args = append(args,
+				"--config",
+				fmt.Sprintf(`core.sshCommand=%s`, strings.Join(sshCmd, " ")),
+			)
+
+		case typeUsernamePassword:
+			repoURL, err := url.Parse(flagValues.url)
+			if err != nil {
+				return err
+			}
+
+			username, err := ioutil.ReadFile(filepath.Join(flagValues.secretPath, "username"))
+			if err != nil {
+				return err
+			}
+
+			password, err := ioutil.ReadFile(filepath.Join(flagValues.secretPath, "password"))
+			if err != nil {
+				return err
+			}
+
+			repoURL.User = url.UserPassword(string(username), string(password))
+
+			credHelperFile, err := ioutil.TempFile(os.TempDir(), "cred-helper-file")
+			if err != nil {
+				return err
+			}
+
+			defer os.Remove(credHelperFile.Name())
+
+			if err := ioutil.WriteFile(credHelperFile.Name(), []byte(repoURL.String()), fs.FileMode(0400)); err != nil {
+				return err
+			}
+
+			if _, err := git(ctx, "config", "--global", "credential.helper", fmt.Sprintf("store --file %s", credHelperFile.Name())); err != nil {
+				return err
+			}
+		}
+	}
+
+	args = append(args, "--", flagValues.url, flagValues.target)
+	_, err := git(ctx, args...)
+	return err
+}
+
+func git(ctx context.Context, args ...string) (string, error) {
+	cmd := exec.CommandContext(ctx, "git", args...)
+	ctxlog.Debug(ctx, cmd.String())
+
+	out, err := cmd.CombinedOutput()
+
+	var output string
+	if out != nil {
+		output = strings.TrimRight(string(out), "\n")
+	}
+
+	if err != nil {
+		// In case the command fails, it is very likely to be an ExitCode error
+		// which contains the exit code of the command. Create a custom exit
+		// error where the command output is placed into the error to be more
+		// readable to the end-user.
+		switch terr := err.(type) {
+		case *exec.ExitError:
+			err = &ExitError{
+				Code:    terr.ExitCode(),
+				Message: output,
+				Cause:   err,
+			}
+		}
+
+		ctxlog.Error(ctx, err, "git command failed",
+			"command", cmd.String(),
+			"output", output,
+		)
+	}
+
+	return output, err
+}
+
+func hasFile(elem ...string) bool {
+	_, err := os.Stat(filepath.Join(elem...))
+	return !os.IsNotExist(err)
+}
+
+func checkCredentials() (credentialType, error) {
+	hasPrivateKey := hasFile(flagValues.secretPath, "ssh-privatekey")
+	isSSHGitURL := sshGitURLRegEx.MatchString(flagValues.url)
+	switch {
+	case hasPrivateKey && !isSSHGitURL:
+		return typeUndef, &ExitError{Code: 110, Message: "Credential/URL inconsistency: SSH credentials provided, but URL is not a SSH Git URL"}
```

Comment:
This is different to the current behavior where there is "only" a warning. Our current e2e tests also add the credential to all tests ignoring if it is to test a private repo or not if I remember correctly.

Reply:
Maybe we can have a call about this. I took the liberty to change the behavior a bit and that we then can discuss the pros and cons.